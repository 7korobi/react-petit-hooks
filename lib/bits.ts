const BITLIMIT = 31
const BITMASK = 2 ** 31 - 1

type Labels<T extends readonly string[]> = T[number][]
type BitsDic<T extends readonly string[], X> = {
  [key in T[number]]: X
}

export class BitsData<T extends readonly string[]> extends Array {
  0: number
  1: BitsShadow<T>
  constructor(value: number, shadow: BitsShadow<T>) {
    super(2)
    this[0] = value
    this[1] = shadow
  }
}

export class BitsShadow<T extends readonly string[]> {
  labels: T
  mask: number
  posi: BitsDic<T, number>
  nega: BitsDic<T, number>
  idx: BitsDic<T, number>

  constructor(labels: T) {
    if (BITLIMIT < labels.length) {
      throw new Error('too much bits.')
    }
    this.labels = labels
    this.mask = 0
    this.posi = {} as any
    this.nega = {} as any
    this.idx = {} as any
    labels.forEach((label: T[number]) => {
      this.posi[label] = this.idx[label] = 0
      this.nega[label] = BITMASK
    })
    labels.forEach((label: T[number], idx) => {
      const posi = 2 ** idx
      const nega = BITMASK & ~posi
      this.mask |= posi
      this.posi[label] |= posi
      this.nega[label] &= nega
      this.idx[label] || (this.idx[label] = idx)
    })
  }

  by(src: T | Labels<T>): number
  by(src: number): Labels<T>
  by(src: T | Labels<T> | number): Labels<T> | number {
    if ('number' === typeof src) {
      const labels: Labels<T> = []
      this.labels.forEach((label: T[number]) => {
        if (src & this.posi[label]) {
          labels.push(label)
        }
      })
      return labels
    } else {
      let n = 0
      src.forEach((label: T[number]) => {
        n |= (this.posi[label] || 0) as number
      })
      return n
    }
  }

  is(n: number): [BitsData<T>, BitsDic<T, number>] {
    const o: BitsData<T> = new BitsData<T>(n, this)
    const proxy = new Proxy(o, {
      get([value, shadow], label: T[number]) {
        return value & shadow.posi[label]
      },
      set(data, label: T[number], set: number | boolean | null) {
        const [value, shadow] = data
        let bits: number
        switch (set) {
          case null:
          case false:
            bits = 0
            break
          case true:
            bits = shadow.posi[label]
            break
          default:
            bits = shadow.posi[label] & (set << shadow.idx[label])
        }
        data[0] = value & shadow.nega[label] | bits
        return true
      },
      has([value, shadow], label: T[number]) {
        return !!shadow.idx[label]
      },
    }) as any
    return [o, proxy]
  }

  isOneBit(x: number) {
    return 0 === (x & (x - 1))
  }

  firstOut(x: number) {
    return x & (x - 1)
  }
  firstIn(x: number) {
    return x | (x + 1)
  }

  firstLinksOut(x: number) {
    return ((x | (x - 1)) + 1) & x
  }
  firstLinksIn(x: number) {
    return ((x & (x + 1)) - 1) | x
  }

  findBitOn(x: number) {
    return x & -x
  }
  findBitOff(x: number) {
    return ~x & (x + 1)
  }

  fillHeadsToOn(x: number) {
    return x | (x - 1)
  }

  fillHeadsToOff(x: number) {
    return x & (x + 1)
  }

  headsBitOff(x: number) {
    return ~x & (x - 1)
  }
  headsBitOn(x: number) {
    return ~(~x | (x + 1))
  }

  headsBitOffAndNextOn(x: number) {
    return x ^ (x - 1)
  }

  snoob(x: number) {
    const minbit = x & -x
    const ripple = x + minbit
    const ones = ((x ^ ripple) >>> 2) / minbit
    return ripple | ones
  }

  count(x: number) {
    let n
    n = (x >>> 1) & 0x77777777
    x = x - n
    n = (n >>> 1) & 0x77777777
    x = x - n
    n = (n >>> 1) & 0x77777777
    x = x - n
    x = (x + (x >>> 4)) & 0x0f0f0f0f
    x = x * 0x01010101
    return x >>> 24
  }

  reverse(x: number) {

  }
}

export function Bits<T extends readonly string[]>(labels: T) {
  return new BitsShadow(labels)
}
