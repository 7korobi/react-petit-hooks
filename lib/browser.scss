body {
  overscroll-behavior: none;
  pointer-events: none;
  user-select: none;

  box-sizing: border-box;
  margin: 0;
  padding: 0;
}

#safe-area-measure {
  pointer-events: none;
  user-select: none;
  opacity: 0;
  visibility: hidden;

  z-index: -99999;

  width: 1%;
  height: 1%;
  margin-top: env(safe-area-inset-top);
  margin-right: env(safe-area-inset-right);
  margin-bottom: env(safe-area-inset-bottom);
  margin-left: env(safe-area-inset-left);
}