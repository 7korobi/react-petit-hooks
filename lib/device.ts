export const __SPEC__ = 'undefined' === typeof window
export const __BROWSER__ = !__SPEC__
export const __TEST__ = __SPEC__ || 'localhost' === location.hostname
