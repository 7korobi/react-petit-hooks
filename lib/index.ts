export * from './browser'
export * from './poll'
export * from './storage'
export * from './time'

