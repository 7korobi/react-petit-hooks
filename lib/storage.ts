import { createContext, useContext, useState, useEffect } from "react"
import * as _ from "lodash"

type UrlData = {
  PATH: string[]
  HASH: [string]
  BASIC_AUTH: [string, string]
  [key: string]: string[]
}
interface Storage<T> {
  getItem(label: string): T;
  setItem(label: string, val: T): void;
  removeItem(label: string): void;
}

const BITLIMIT = 31
const BITMASK = 2**31-1
export class Bits {
  static by_str(o: string, base: any) {
    const bits = [null, undefined, "undefined"].includes(o) ? 0 : Number.parseInt(o, 36)
    return base._.copy(bits)
  }
  static to_str(o: any) {
    return o.bits.toString(36)
  }

  static min(x: number){ return x & (-x) }

  static assign<L extends string>(labels: readonly L[] ) {
    if (BITLIMIT < labels.length) { throw new Error("too much bits.") }

    class BitsProcess {
      labels: readonly L[] = labels

      constructor (public target: BitsBox) {}
      copy(x = this.target.bits): BitsBox {
        return new BitsBox(x) 
      }
      calc(cb: (x:number) => number){
        const x = this.target.bits
        return this.copy(cb(x))
      }
      get labels_off(){ return BitsProcess.get_labels(BITMASK - this.target.bits) }
      get labels_on(){ return BitsProcess.get_labels(this.target.bits) }

      static get_labels(x){
        let count = labels.length
        let idx = 0
        const res: L[] = []
        do {
          if (x & 1 && labels[idx]) {
            res.push(labels[idx]) 
          }
          x >>>= 1
          ++idx
        } while( --count )
        return res.reverse()
      }
    }
    
    class BitsBox extends Bits {
      _: BitsProcess
      constructor(public bits: number) {
        super()
        this._ = new BitsProcess(this)
      }
      static all = new BitsBox(BITMASK)
      static zero = new BitsBox(0)
      static by(labels: L[]): BitsBox {
        const o: any = new BitsBox(0)
        labels.forEach((label)=>{
          o[label] = true
        })
        return o
      }
    }

    labels.forEach((label, idx)=>{
      const bit = 2**idx
      const mask = BITMASK - bit
      Object.defineProperty(BitsBox.prototype, label,{
        enumerable: true,
        get: function (this: BitsBox): boolean { return Boolean(this.bits & bit) },
        set: function (this: BitsBox, b: boolean ){
          this.bits = this.bits & mask | (bit * Number(to_Boolean(b))) 
        }
      })
    })
    return BitsBox
  }
}

//
// format.
//
function to_String(u: any, nil: ''): string
function to_String(u: any, nil: undefined): string | undefined
function to_String(u: any, nil){ return u ? String(u) : nil }
function to_Number(u: any): number { return u ? Number(u) : NaN }
function to_Array(u: any): any[] { return u instanceof Array ? u : u ? Array(u) : [] }
function to_Object(u: any): object { return u instanceof Object ? u : {} }
function to_Boolean(u: any): boolean { return !!u && ! ['0', 'false'].includes(u) }

type VALUE = string | number | boolean | Bits
type VALUE_SET = VALUE | VALUE[]
type VALUE_SETS = {[key: string]: VALUE_SET }
type VALUE_TREE = VALUE_SET | VALUE_TREES
type VALUE_TREES = {[key: string]: VALUE_TREE }

function to_str<T extends VALUE_TREE>(o: T, base: T): string
function to_str(o, base) {
  switch (base.constructor) {
    case String  : return to_String(o, '')
    case Number  : return to_String(o, '')
    case Boolean : return to_String(to_Boolean(o), '')
  }
  if ( base instanceof Bits ) { return Bits.to_str(o) }
  if ( base instanceof Array ) { return JSON.stringify(to_Array(o)) }
  return JSON.stringify(to_Object(o))
}

function by_str<T extends VALUE_TREE>(o: string | null, base: T): T
function by_str(o, base) {
    switch (base.constructor) {
    case String  : return to_String(o, undefined)
    case Number  : return to_Number(o)
    case Boolean : return to_Boolean(o)
  }
  if ( base instanceof Bits ) { return Bits.by_str(o, base) }
  if ( base instanceof Array ) { return JSON.parse(o) || [] }
  return JSON.parse(o) || {}
}

function to_url<T extends VALUE_SET>(o: T, base: T): string[]
function to_url(o, base) {
  switch (base.constructor) {
    case String  : return [encodeURIComponent( to_String(o, '') )]
    case Number  : return [to_String(to_Number(o), '')]
    case Boolean : return [to_String(to_Boolean(o), '')]
  }
  if ( base instanceof Bits ) { return [Bits.to_str(o)] }
  if ( base instanceof Array ) { return to_Array(o).map(encodeURIComponent) }
}

function by_url<T extends VALUE_SET>(o: string[], base: T): T
function by_url(o, base) {
    switch (base.constructor) {
    case String  : return to_String(decodeURIComponent(o[0]), '')
    case Number  : return to_Number(decodeURIComponent(o[0]))
    case Boolean : return to_Boolean(o[0])
  }
  if ( base instanceof Bits ) { return Bits.by_str(o[0], base) }
  if ( base instanceof Array ) { return to_Array(o).map(decodeURIComponent) }
}


//
// data store.
//
const defaults = {}
const dataStore = {}
const share = {}
export const debug = { share, dataStore, defaults }

function doShare(path: string[], tail, val, base) {
  const key = path.join('.')
  if (! share[key] ) { return }
  share[key].forEach((cb) => {
    cb(val, base)
  })
  // console.warn( path, tail, share[key].length )
}

function doSharesDown(path: string[], tail, val, base) {
  if ( base instanceof Object ) {
    Object.keys(val).forEach((key)=>{
      if ( tail === key ) { return }
      if ('_' === key[0]) { return }
      const extPath = [...path, key]
      const base = _.get(defaults, extPath)
      const val = _.get(dataStore, extPath)
      doSharesDown(extPath, undefined, val, base)
    })
  }
  doShare(path, tail, val, base)
}

function doSharesUp(path: string[]) {
  let idx = path.length
  do {
    const tail = path[idx]
    const subPath = path.slice(0, idx)
    const base = _.get(defaults, subPath)
    const val = _.get(dataStore, subPath)
    doSharesDown(subPath, tail, val, base)
  } while (--idx)
}

export function pushState(o: VALUE_SETS) { defineUrlStore(pushStateStorage, o) }
export function replaceState(o: VALUE_SETS) { defineUrlStore(replaceStateStorage, o) }
function defineUrlStore(store: UrlStorage, o: VALUE_SETS){
  defineStore<VALUE_SET,VALUE_SETS,string[]>(store, o, to_url, by_url)
  store.reset(o)
}
export function localStore(o) { defineStore<VALUE_TREE,VALUE_TREES,string | null>(LocalStorage, o, to_str, by_str) }
export function sessionStore(o) { defineStore<VALUE_TREE,VALUE_TREES,string | null>(SessionStorage, o, to_str, by_str) }
function defineStore<T,TT,P>(storage: Storage<P>, o: TT, to: (o: T, base: T) => P, by: (o: P, base: T) => T) {
  Object.keys(o).forEach((rootPath)=>{
    const val = o[rootPath]
    const item = storage.getItem(rootPath)
    const rootCb = _.debounce((val) => {
      if (val) {
        storage.setItem(rootPath, to(val, defaults[rootPath]))
      } else {
        storage.removeItem(rootPath)
      }
    }, 100);

    defaults[rootPath] = val
    if (!share[rootPath]) {
      share[rootPath] = new Set()
    }
    share[rootPath].add(rootCb)

    if (item) {
      dataStore[rootPath] = by(item, val)
    } else {
      dataStore[rootPath] = val
      rootCb(val)
    }
  })
}

class UrlStorage {
  rootPaths: string[]
  data: UrlData

  getItem(key: string): string[] {
    return this.data[key]
  }

  setItem(key: string, val: string[]) {
    if (val !== this.data[key]) {
      this.data[key] = val
      this.store()
    } else {
      this.data[key] = val
    }
  }

  removeItem(key: string) {
    if (this.data[key]) {
      delete this.data[key]
      this.store()
    } else {
      delete this.data[key]
    }
  }

  constructor(public mode: 'pushState' | 'replaceState') {
    this.rootPaths = []
    this.data = {PATH:[""], HASH:[""], BASIC_AUTH: ["", ""]}
  }

  reset(defs: VALUE_SETS){
    this.rootPaths.forEach((rootPath)=>{
      rootPath
    })
    this.rootPaths = Object.keys(defs)
    this.data = UrlStorage.parse()
    this.store('replaceState')
  }

  store(mode: 'pushState' | 'replaceState' = this.mode) {
    const href = UrlStorage.stringify(this.data)
    history[mode](this.data, '', href)
  }


  static parse(href = BaseUrl(), base = BaseUrl()): UrlData {
    const { hash, search, pathname, username, password, protocol, host, hostname } = new URL(href, base)
    const data: UrlData = {
      PATH: pathname.split(/\/+/).slice(1),
      HASH: [hash.slice(1)],
      BASIC_AUTH: [username, password],
    }
    search.slice(1).split('&').forEach((query) => {
      const [key, ...vals] = query.split("=")
      if (!key) { return }
      data[key] = vals
    })
    return data
  }

  static stringify(data: UrlData, base = BaseUrl()): string {
    const hash = data.HASH[0] ? `#${data.HASH[0]}` : ''
    let search = ""
    for (const key in data) {
      switch (key) {
        case 'PATH':
        case 'HASH':
        case 'BASIC_AUTH':
          break;
        default:
          if (data[key].length) {
            search += `&${key}=${data[key].join("=")}`
          } else {
            search += `&${key}`
          }
      }
    }
    if (search) {
      search = `?${search.slice(1)}`
    }
    return new URL(`${search}${hash}`, base).href
  }
}
export const pushStateStorage = new UrlStorage('pushState')
export const replaceStateStorage = new UrlStorage('replaceState')


export function useStore<T>(path_obj: any): [T, (val: T) => void] {
  const path = _.toPath(path_obj)
  const key = path.join('.')
  const setter = (val: T) => {
    _.set(dataStore, path, val)
    doSharesUp(path)
  }
  if (!share[key]) {
    share[key] = new Set()
  }
  const [item, setItem] = useState<T>(_.get(dataStore, path))
  useEffect(() => {
    share[key].add(setItem)
    return () => {
      share[key].delete(setItem)
    }
  }, [key])
  return [item, setter]
}

let BaseUrl: ()=> string
let LocalStorage : Storage<string | null>
let SessionStorage : Storage<string | null>

if (typeof window !== "undefined") {
  BaseUrl = () => window.location.href
  LocalStorage = window.localStorage
  SessionStorage = window.sessionStorage
  window.addEventListener('popstate', (e) => {
    const data = UrlStorage.parse()

    console.warn(e)
  })
  window.addEventListener("storage", ({ key, newValue }) => {
    if (!key || !newValue) { return }
    const base = defaults[key]
    if (!base) { return }
    const val = by_str(newValue, base)
    dataStore[key] = val
    doSharesUp([key])
  })

} else {
  BaseUrl = () => "https://localhost/"
  const dummyStorage : Storage<string | null> = {
    getItem(label) { return '' },
    setItem(label) {},
    removeItem(label) {}
  }
  LocalStorage = SessionStorage = dummyStorage
}
