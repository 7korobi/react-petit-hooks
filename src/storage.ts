import { useState, useEffect } from 'react'

import { BitsData } from './bits'
import { __BROWSER__ } from './device'

export function useLocalStorage<T>(key: string, base: T) {
  return useStorage('localStorage', key, base)
}

export function useSessionStorage<T>(key: string, base: T) {
  return useStorage('localStorage', key, base)
}

function to_String(u: any, nil: ''): string
function to_String(u: any, nil: undefined): string | undefined
function to_String(u: any, nil: undefined | '') {
  return u ? String(u) : nil
}
function to_Number(u: any): number {
  return u ? Number(u) : 0 === u ? 0 : NaN
}
function to_Array(u: any): any[] {
  return u instanceof Array ? u : u ? Array(u) : []
}
function to_Object(u: any): object {
  return u instanceof Object ? u : {}
}
function to_Boolean(u: any): boolean {
  return !!u && !['0', 'false'].includes(u)
}

function to_str(o: any, base: any): string {
  if ('string' === typeof base) {
    return to_String(o, '')
  }
  if ('number' === typeof base) {
    return to_String(o, '')
  }
  if ('boolean' === typeof base) {
    return to_String(to_Boolean(o), '')
  }
  if (base instanceof BitsData) {
    return base.field.to_str(o as any)
  }
  if (base instanceof Array) {
    return JSON.stringify(to_Array(o).map((arg, idx) => to_str(arg, base[idx] || base[0])))
  }
  if (base instanceof Object) {
    const val: { [key: string]: any } = Object.assign({}, to_Object(o))
    for (const key in base) {
      val[key] = to_str(val[key], base[key])
    }
    return JSON.stringify(val)
  }
  throw new Error(`bad data. ${o} ${base}`)
}

function by_str(o: string | null | undefined, base: any): any {
  if (undefined === o) {
    return base
  }
  if ('string' === typeof base) {
    return to_String(o, undefined)
  }
  if ('number' === typeof base) {
    return to_Number(o)
  }
  if ('boolean' === typeof base) {
    return to_Boolean(o)
  }
  if (base instanceof BitsData) {
    return base.field.by_str(o as any)
  }
  if (base instanceof Array) {
    return ((o ? JSON.parse(o) || [] : []) as string[]).map((arg, idx) =>
      by_str(arg, base[idx] || base[0])
    )
  }
  if (base instanceof Object) {
    const val: { [key: string]: any } = o ? JSON.parse(o) || {} : {}
    for (const key in base) {
      val[key] = by_str(val[key], base[key])
    }
    return val
  }
  throw new Error(`bad data. ${o} ${base}`)
}

function useStorage<T>(
  storage: 'localStorage' | 'sessionStorage',
  key: string,
  base: T
): [T, (data: T) => void] {
  const [data, setData] = useState(base)
  useEffect(init, [])
  useEffect(refresh, [data])

  return [data, setData]

  function init() {
    const data = by_str(window[storage].getItem(key) || undefined, base)
    __BROWSER__ && setData(data)

    if ('sessionStorage' === storage) {
      return
    }
    window.addEventListener('storage', onStorage)
    return () => {
      window.removeEventListener('storage', onStorage)
    }
  }
  function refresh() {
    __BROWSER__ && window[storage].setItem(key, to_str(data, base))
  }
  function onStorage(e: StorageEvent) {
    if (e.key !== key) {
      return
    }
    const data = by_str(e.newValue, base)
    setData(data)

    if (e.oldValue === null && e.newValue !== null) {
      console.log('type : add')
    } else if (e.oldValue !== null && e.newValue === null) {
      console.log('type : remove')
    } else if (e.oldValue !== null && e.newValue !== null) {
      console.log('type : update')
    } else if (e.oldValue === null && e.newValue === null && e.storageArea!.length === 0) {
      console.log('type : clear')
    }
  }
}

function getUrl() {
  return new URL(__BROWSER__ ? location.href : 'https://localhost/')
}

function getUrlData(): [URL, { [key: string]: string[] }, string[]] {
  const url = getUrl()
  const urlHash = url.hash.slice(1).split('=').map(decodeURIComponent)
  const urlSearch = {}
  for (const str of url.search.slice(1).split('&')) {
    if (!str) {
      continue
    }
    const [key, ...val] = str.split('=').map(decodeURIComponent)
    if (!urlSearch[key]) {
      urlSearch[key] = []
    }
    urlSearch[key].push(...val)
  }
  return [url, urlSearch, urlHash]
}

function to_url(o: any, base: any): string[] {
  if ('string' === typeof base) {
    return [to_String(o, '')]
  }
  if ('number' === typeof base) {
    return [to_String(to_Number(o), '')]
  }
  if ('boolean' === typeof base) {
    return [to_String(to_Boolean(o), '')]
  }
  if (base instanceof BitsData) {
    return [base.field.to_str(o as any)]
  }
  if (base instanceof Array) {
    return to_Array(o).map((arg, idx) => to_url(arg, base[idx] || base[0])[0])
  }
  throw new Error(`bad data. ${o} ${base}`)
}

function by_url(args: string[], base: any): any {
  if (!args.length) {
    return base
  }
  if ('string' === typeof base) {
    return to_String(args[0], '')
  }
  if ('number' === typeof base) {
    return to_Number(args[0])
  }
  if ('boolean' === typeof base) {
    return to_Boolean(args[0])
  }
  if (base instanceof BitsData) {
    return base.field.by_str(args[0] as any)
  }
  if (base instanceof Array) {
    return args.map((arg, idx) => by_url([arg], base[idx] || base[0]))
  }
  throw new Error(`bad data. ${args} ${base}`)
}

export function useUrlState<Q, H>(
  mode: 'pushState' | 'replaceState',
  search: Q,
  hash?: H
): [[Q] | [Q, H], (data: [Q] | [Q, H]) => void] {
  const [data, setData] = useState<[Q] | [Q, H]>(hash ? [search, hash] : [search])
  useEffect(init, [])

  return [data, setUrl]

  function init() {
    onChange()
    window.addEventListener('popstate', popup)
    return () => {
      window.removeEventListener('popstate', popup)
    }
  }

  function onChange() {
    const [url, urlSearch, urlHash] = getUrlData()
    const hashVal = capture(urlHash, hash)
    const searchVal: Q = {} as Q
    for (const key in search) {
      searchVal[key] = capture(urlSearch[key], search[key])
    }
    setData([searchVal, hashVal])
  }

  function capture(data, base) {
    if (data) {
      if (data[0].length) {
        return by_url(data, base)
      }
    }
    return by_url([], base)
  }

  function popup(e: PopStateEvent) {
    console.log(e, e.state)
    onChange()
  }

  function setUrl(data: [Q] | [Q, H]): void {
    let [url, urlSearch, urlHash] = getUrlData()

    if (data[1]) {
      url.hash = to_url(data[1], hash).map(encodeURIComponent).join('=')
    }

    for (const key in data[0]) {
      urlSearch[key] = to_url(data[0][key], search[key])
    }

    const searchStrings: string[] = []
    for (const key in urlSearch) {
      searchStrings.push(
        `${encodeURIComponent(key)}=${urlSearch[key].map(encodeURIComponent).join('=')}`
      )
    }
    url.search = searchStrings.join('&')

    __BROWSER__ && history[mode](data, '', url.href)
    onChange()
  }
}
