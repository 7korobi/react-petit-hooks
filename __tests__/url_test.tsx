import { render, waitFor, act } from '@testing-library/react'
import '@testing-library/jest-dom/extend-expect'

import React from 'react'

import { Bits } from '../lib/bits'
import { useLocalStorage } from '../lib/storage'

const TestBits = new Bits(['a', 'b', 'c', 'd', 'e', 'f', 'g'] as const)
const data = {
  data_b: false,
  data_s: 'test string data',
  data_n: 123456789,
  bits: TestBits.data(TestBits.by(['a', 'c', 'e'])),
}

function TestData() {
  const [test, setTest] = useLocalStorage('test_key', data)
  return (
    <p>
      <b>{test.data_b ? '〇' : '✖'}</b>
      <b>{test.data_s}</b>
      <b>{test.data_n}</b>
      <b>{test.bits.is.a ? '〇' : '✖'}</b>
      <b>{test.bits.is.b ? '〇' : '✖'}</b>
      <b>{test.bits.is.c ? '〇' : '✖'}</b>
      <b>{test.bits.is.d ? '〇' : '✖'}</b>
      <b>{test.bits.is.e ? '〇' : '✖'}</b>
      <b>{test.bits.is.f ? '〇' : '✖'}</b>
      <b>{test.bits.is.g ? '〇' : '✖'}</b>
    </p>
  )
}
test('basic value', () => {
  const c = render(<TestData />)
  expect(c.container).toMatchSnapshot()
  c.rerender(<TestData />)
  expect(c.container).toMatchSnapshot()
  c.unmount()
  expect(c.container).toMatchSnapshot()
  expect(data).toMatchSnapshot()
})
