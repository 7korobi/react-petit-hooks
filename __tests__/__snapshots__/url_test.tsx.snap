// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`basic value 1`] = `
<div>
  <p>
    <b>
      ✖
    </b>
    <b>
      test string data
    </b>
    <b>
      123456789
    </b>
    <b>
      〇
    </b>
    <b>
      ✖
    </b>
    <b>
      〇
    </b>
    <b>
      ✖
    </b>
    <b>
      〇
    </b>
    <b>
      ✖
    </b>
    <b>
      ✖
    </b>
    <b>
      http://localhost/
    </b>
  </p>
</div>
`;

exports[`basic value 2`] = `
<div>
  <p>
    <b>
      ✖
    </b>
    <b>
      test string data
    </b>
    <b>
      123456789
    </b>
    <b>
      〇
    </b>
    <b>
      ✖
    </b>
    <b>
      〇
    </b>
    <b>
      ✖
    </b>
    <b>
      〇
    </b>
    <b>
      ✖
    </b>
    <b>
      ✖
    </b>
    <b>
      http://localhost/?data_b=&data_s=test%20string%20data&data_n=123456789&bits=l
    </b>
  </p>
</div>
`;

exports[`basic value 3`] = `<div />`;

exports[`basic value 4`] = `
Object {
  "bits": BitsData {
    "field": Bits {
      "idx": Object {
        "a": 0,
        "b": 1,
        "c": 2,
        "d": 3,
        "e": 4,
        "f": 5,
        "g": 6,
      },
      "labels": Array [
        "a",
        "b",
        "c",
        "d",
        "e",
        "f",
        "g",
      ],
      "mask": 127,
      "nega": Object {
        "a": 2147483646,
        "b": 2147483645,
        "c": 2147483643,
        "d": 2147483639,
        "e": 2147483631,
        "f": 2147483615,
        "g": 2147483583,
      },
      "posi": Object {
        "a": 1,
        "b": 2,
        "c": 4,
        "d": 8,
        "e": 16,
        "f": 32,
        "g": 64,
      },
    },
    "has": Object {
      "field": 0,
      "has": 0,
      "is": 0,
      "value": 0,
    },
    "is": Object {
      "field": false,
      "has": false,
      "is": false,
      "value": false,
    },
    "value": 21,
  },
  "data_b": false,
  "data_n": 123456789,
  "data_s": "test string data",
}
`;
